util/EIP-712/src/encode.rs

```diff
@@ -0,0 +1,339 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! EIP712 Encoder
+use ethabi::{encode, Token as EthAbiToken};
+use ethereum_types::{Address as EthAddress, U256, H256};
+use keccak_hash::keccak;
+use serde_json::Value;
+use std::str::FromStr;
+use itertools::Itertools;
+use linked_hash_set::LinkedHashSet;
+use serde_json::to_value;
+use parser::{Parser, Type};
+use error::{Result, ErrorKind, serde_error};
+use eip712::{EIP712, MessageTypes};
+
+/// given a type and HashMap<String, Vec<FieldType>>
+/// returns a HashSet of dependent types of the given type
+fn build_dependencies<'a>(message_type: &'a str, message_types: &'a MessageTypes) -> Option<(LinkedHashSet<&'a str>)>
```

Comment:
My previous question is still valid: Can you argument why we need `LinkedHashSet` here?

Reply:
`std::collections::HashSet` doesn't have a `pop` method, ``LinkedHashSet` has a `pop_back` method,

Reply:
Still not convinced:
1. we use `pop_back` for `types` not `deps` (the return type)
2. Why do we need to process `types` in order?

Reply:
To answer "why does `types` have to be  `HashSet`"?, It's an optimization.

I could either 
1. iterate through the `types` to assert that the `type` i'm adding to the array doesn't already exist.
2. I could just use a hashset.

Reply:
@seunlanlege I was asking more about `Linked` variant of `HashSet` (which requires an external crate). I'm completely ok with how it is implemented right now, just trying to understand the design decision behind `LinkedHashSet`, like:
1. Is it required by the spec (the order of items matter)
2. Is it required for tests (makes asserting that something is on particular position in the hashset)

The minimalist in me is just reluctant to include an external crate if it's not strictly required :)

Reply:
Actually no the spec doesn't in any way require a `LinkedHashSet`, I require a `HashSet` that i can `pop` from; which the one in the stdlib lacks.

Reply:
I can make `deps` `std::collection::HashSet`; it doesn't actually need to be `LinkedHashSet`